### PR TITLE
Fixed a few ip issues

### LINKED ISSUE #1175: EC2 Instances in VPC do not get IPs that match the VPC Subnet
When creating an EC2 instances inside of a subnet with moto, the IP addresses that the instance gets assigned does not align with the subnet that the instance is being launched in.  Additionally, when setting the subnet attribute `MapPublicIpOnLaunch` to `True`, the value is not respected when the instance is launched in said subnet. 

Here's a script to easily reproduce:

```
import boto3, botocore, moto, sys
from moto import mock_ec2

@mock_ec2
def main():
    print('python version:', sys.version)
    print('moto version:', moto.__version__)
    print('boto3 version:', boto3.__version__)
    print('botocore version:', botocore.__version__)
    client = boto3.client('ec2', 'us-east-1')

    result = client.create_vpc(CidrBlock='192.168.0.0/16')
    vpc_id = result['Vpc']['VpcId']
    print('Created VpcId=', vpc_id)

    result = client.create_subnet(CidrBlock='192.168.1.0/24', VpcId=vpc_id)
    subnet_id = result['Subnet']['SubnetId']
    print('Created SubnetId=', subnet_id)

    client.modify_subnet_attribute(
        SubnetId=subnet_id,
        MapPublicIpOnLaunch={'Value': True}
    )

    resource = boto3.resource('ec2', 'us-east-1')
    i = resource.create_instances(
        ImageId='ami-abc123',
        MinCount=1,
        MaxCount=1,
        SubnetId=subnet_id
    )[0]

    print('instance_id=', i.instance_id)
    print('private_ip_address=', i.private_ip_address)
    print('public_ip_address=', i.public_ip_address)
    print('subnet_id=', i.subnet_id)
    print('subnet.cidr_block=', i.subnet.cidr_block)
    print('vpc_id=', i.vpc_id)
    print('vpc.cidr_block=', i.vpc.cidr_block)

if __name__ == '__main__':
    exit(main())
```

The output I got when running:

```
python version: 3.5.2 (v3.5.2:4def2a2901a5, Jun 26 2016, 10:47:25) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
moto version: 1.0.1
boto3 version: 1.4.4
botocore version: 1.5.73
Created VpcId= vpc-3679181b
Created SubnetId= subnet-36678a48
instance_id= i-9542d363
private_ip_address= 10.86.170.112
public_ip_address= None
subnet_id= subnet-36678a48
subnet.cidr_block= 192.168.1.0/24
vpc_id= vpc-3679181b
vpc.cidr_block= 192.168.1.0/16
```

What I would expect:

```
 ... 
private_ip_address= 192.168.1.X
public_ip_address= X.X.X.X
 ...
```

I expect `private_ip_address` to return a private ip address within the subnet address range (`192.168.1.0/24` in this example) and `public_ip_address` to return a public ip address address vs `None`.  Additionally, it needs to take into consideration [reserved IPs](http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Client.create_subnet) by AWS. 

This is my first time using moto (nice job btw!) so I'm not familiar with the code base.  I'm happy to submit a PR if you can point me in the direction of where the edits would need to take place.

### PR DESCRIPTION
**Pull request by terrycain**
Ok I changed how we assign IPs to instances so that they get assigned IPs that fall within the subnet the instance is started in.

Fixes #1175 

Travis has spammed many errors at me, am sure most of them are bs, will look into it in the morning.

### CODE REVIEW

File: moto/rds2/models.py [hunk 1]
```diff
@@ -704,7 +704,8 @@ def create_database_replica(self, db_kwargs):
         if self.arn_regex.match(source_database_id):
             db_kwargs['region'] = self.region
 
-        replica = copy.deepcopy(primary)
+        # Shoudln't really copy here as the instance is duplicated. RDS replicas have different instances.
```
Comment (JackDanger) [inline moto/rds2/models.py]: Typo on the comment here. "Shoudln't" -> "Shouldn't"

Comment (terrycain) [reply moto/rds2/models.py]: Haha :smile:

**Comment by JackDanger** [comment]
Travis is having networking issues downloading images from Dockerhub but once that's clear this should be green.

**Comment by terrycain** [comment]
Ive Added in 5 minute timeouts as well, as the lambda should be killed then anyway. Also async lambda test was referencing wrong handler function.

**Comment by terrycain** [comment]
I also upped the timeout for starting moto, I can't see why it would take longer though

**Comment by terrycain** [comment]
Meh, until travis fixes itself, I deem the PR done.

**Comment by terrycain** [comment]
@JackDanger It passes :D:D, review plz

File: moto/awslambda/models.py [hunk 2]
```diff
@@ -298,7 +298,12 @@ def _invoke_lambda(self, code, event=None, context=None):
                         volumes=["{}:/var/task".format(data_vol.name)], environment=env_vars, detach=True, **run_kwargs)
                 finally:
                     if container:
-                        exit_code = container.wait()
+                        try:
+                            exit_code = container.wait(timeout=300)
```
Comment (JackDanger) [inline moto/awslambda/models.py]: Thanks for this!

**Pull request merged**